File: backend/src/edgereco/reco/reranker.py
"""Rerank search results using session profile, behind an absolute relevance floor.

THE FLOOR EXISTS BECAUSE RANKING ALONE CANNOT SAY "NO"
RRF fuses POSITIONS, not scores: a document ranked first by one retriever scores
1/61 whether it is a perfect match or the least bad of 720 wrong answers. The
rank-normalization below then divides by the best hit in this very result set, so
the top result takes the full relevance weight for every query ever asked. Nothing
downstream could express "no good match", and a full page came back for queries the
catalog cannot answer at all. Demoting junk does not fix that; refusing it does.

Mirrored byte-for-byte in the browser port
(``frontend/packages/edgeproc-browser/src/engine/reranker.ts``); a test in
``tests/unit/reco/test_reranker.py`` reads that file and pins both constants,
because one rule written in two languages will otherwise diverge.
"""

from __future__ import annotations

from typing import Final, NamedTuple

from edgereco.catalog.models import SearchResult, SessionProfile
from edgereco.reco.ranking_config import DEFAULT_RANKING_CONFIG, ScoringWeights
from edgereco.reco.scorer import score_product

_SEARCH_RELEVANCE_WEIGHT: Final[float] = 0.2

#: The absolute semantic floor: the cosine below which a document is, empirically,
#: not an answer. CALIBRATED, NOT CHOSEN — 0.3970 is the highest cosine any document
#: reaches for a query the catalog cannot answer, measured over the golden set's 8
#: ``negative`` queries (576 candidate observations). The floor is that noise ceiling
#: rounded up to the next hundredth.
MIN_SEMANTIC_RELEVANCE: Final[float] = 0.4

#: The absolute lexical floor: any strictly-positive BM25 score is evidence. Same
#: rule, same 8 queries, which produced ZERO positive BM25 scores — the measured
#: lexical noise ceiling is 0. A hit means the query shares a discriminating term
#: with the product's indexed text, which is absolute evidence whatever the
#: embedding thinks. Honest limitation: those queries are nonsense words by
#: construction and so can never produce a BM25 hit, making this side of the floor
#: calibrated on a tautology. Dropping the lexical branch is not the safer option —
#: a semantic-only floor at 0.4 silences three real golden-set queries outright.
MIN_LEXICAL_RELEVANCE: Final[float] = 0.0


class RetrievalEvidence(NamedTuple):
    """What each retriever measured for one candidate, on its own absolute scale.

    ``None`` means "this retriever never scored it", which is not the same as 0.0.
    RRF and the rank-normalization both erase these magnitudes by design, so this is
    the only place they survive to reach the floor.
    """

    semantic: float | None
    """Cosine to the query embedding; None when the vector index did not return it."""

    lexical: float | None
    """BM25 score; None when the keyword index did not return it."""


def meets_relevance_floor(evidence: RetrievalEvidence | None) -> bool:
    """Does this candidate clear the floor on EITHER retriever?

    Hybrid retrieval produces two independent absolute signals and a document needs
    only one of them to be a defensible answer. Missing evidence is a refusal, not a
    pass: a candidate the caller cannot account for is dropped.
    """
    if evidence is None:
        return False
    if evidence.semantic is not None and evidence.semantic >= MIN_SEMANTIC_RELEVANCE:
        return True
    return evidence.lexical is not None and evidence.lexical > MIN_LEXICAL_RELEVANCE


def retrieval_evidence(
    keyword_hits: list[tuple[str, float]],
    vector_hits: list[tuple[str, float]],
) -> dict[str, RetrievalEvidence]:
    """Both retrievers' raw scores keyed by product id — the fusion input, unfused."""
    lexical = dict(keyword_hits)
    semantic = dict(vector_hits)
    return {
        pid: RetrievalEvidence(semantic=semantic.get(pid), lexical=lexical.get(pid))
        for pid in {**lexical, **semantic}
    }


def _retrieval_score(score: float, maximum: float) -> float:
    """Normalize non-negative RRF into the query-relevance component."""
    if maximum <= 0.0:
        return 0.0
    return _SEARCH_RELEVANCE_WEIGHT * max(0.0, score) / maximum


def rerank(
    results: list[SearchResult],
    profile: SessionProfile,
    weights: ScoringWeights = DEFAULT_RANKING_CONFIG.scoring_weights,
) -> list[SearchResult]:
    """Personalize an already-selected recommendation candidate pool."""
    return _descending([score_product(r.product, profile, weights) for r in results])


def rerank_search(
    results: list[SearchResult],
    profile: SessionProfile,
    evidence: dict[str, RetrievalEvidence],
    weights: ScoringWeights = DEFAULT_RANKING_CONFIG.scoring_weights,
) -> list[SearchResult]:
    """Drop what fails the absolute floor, then blend retrieval with session signals.

    ``evidence`` is required, not defaulted: a caller that cannot say what the
    retrievers actually measured gets an empty page, never a silent full one.
    """
    admitted = [r for r in results if meets_relevance_floor(evidence.get(r.product.id))]
    maximum = max((result.score for result in admitted), default=0.0)
    rescored = [
        score_product(
            result.product,
            profile,
            weights,
            retrieval=_retrieval_score(result.score, maximum),
        )
        for result in admitted
    ]
    return _descending(rescored)


def _descending(results: list[SearchResult]) -> list[SearchResult]:
    results.sort(key=lambda result: result.score, reverse=True)
    return results

File: frontend/packages/edgeproc-browser/src/engine/reranker.ts
// Session-aware reranker, ported from edge-reco's reco/scorer.py + reco/reranker.py.
// scoreProduct produces a live Assay explanation; rerank re-scores a result set
// and sorts it descending. The weights and ordered formula match scorer.py exactly:
//
//   retrieval + w_popularity*popularity + w_category*category_match
//   + w_tag*tag_match + w_brand*brand_match + w_freshness*freshness
//   + w_similarity*similarity + w_cooccurrence*cooccurrence
//   - w_repetition*was_recently_viewed
//
// tag_match is the MEAN tag affinity over the product's tags (0 if it has none).
// Retrieval is normalized RRF for search and zero for recommendation-only rails.

import type { ScoreResult } from "@edgeproc/assay";
import type { Product, ScoreComponents, SearchResult } from "./domain";
import { explainScore, type FormulaSignals } from "./formula";
import { DEFAULT_RANKING_CONFIG, type ScoringWeights } from "./rankingConfig";
import type { RankedHit } from "./rerank";
import type { SessionProfile } from "./session";

// Search intent is the primary signal. RRF is normalized to [0, 1] before this
// weight is applied, so popularity can refine the fused ranking without erasing it.
const SEARCH_RELEVANCE_WEIGHT = 0.2;

/**
 * What each retriever actually measured for one candidate, on its own scale and
 * independent of every other candidate in the set.
 *
 * `null` means "this retriever never scored it", which is not the same as 0. RRF
 * and the rank-normalization below both erase these numbers by design — RRF keeps
 * only positions, and the normalization divides by the best hit in the set — so a
 * result set of pure noise looks identical to a set of perfect matches once it
 * reaches the ranker. This is the only place the absolute magnitudes survive.
 */
export interface RetrievalEvidence {
	/** Cosine to the query embedding; null when the vector index did not return it. */
	readonly semantic: number | null;
	/** BM25 score; null when the keyword index did not return it. */
	readonly lexical: number | null;
}

/**
 * The absolute semantic floor: the cosine below which a document is, empirically,
 * not an answer.
 *
 * CALIBRATED, NOT CHOSEN. 0.3970 is the highest cosine any document in the catalog
 * reaches for a query the catalog cannot answer — measured over the golden set's 8
 * `negative` queries (`__fixtures__/relevanceGoldenSet.ts`), 576 candidate
 * observations. That is what noise looks like at its loudest, so the floor is that
 * ceiling rounded up to the next hundredth. Everything at or below it is
 * indistinguishable from a query with no answer at all.
 */
export const MIN_SEMANTIC_RELEVANCE = 0.4;

/**
 * The absolute lexical floor: any strictly-positive BM25 score is evidence.
 *
 * Calibrated by the same rule against the same 8 unanswerable queries, which
 * produced ZERO positive BM25 scores across those 576 observations — the measured
 * lexical noise ceiling is 0. A hit means the query literally shares a
 * discriminating term with the product's indexed text, which is absolute evidence
 * no matter what the embedding thinks. Honest limitation: the negative queries are
 * nonsense words by construction, so they can never produce a BM25 hit; this side
 * of the floor is calibrated on a tautology and a wider negative set could tighten
 * it. Dropping the lexical branch is not the safer option — a semantic-only floor
 * at 0.4 silences three real queries in the golden set outright.
 */
export const MIN_LEXICAL_RELEVANCE = 0;

/**
 * Does this candidate clear the floor on EITHER retriever?
 *
 * Hybrid retrieval produces two independent absolute signals and a document needs
 * only one of them to be a defensible answer. Missing evidence is a refusal, not a
 * pass: a candidate the caller cannot account for is dropped.
 */
export function meetsRelevanceFloor(
	evidence: RetrievalEvidence | undefined,
): boolean {
	if (evidence === undefined) {
		return false;
	}
	const semantic = evidence.semantic;
	if (semantic !== null && semantic >= MIN_SEMANTIC_RELEVANCE) {
		return true;
	}
	const lexical = evidence.lexical;
	return lexical !== null && lexical > MIN_LEXICAL_RELEVANCE;
}

/**
 * Both retrievers' raw scores, keyed by product id — the fusion input before RRF
 * throws the magnitudes away. Absent on one side stays `null`, never 0.
 */
export function retrievalEvidence(
	keywordHits: ReadonlyArray<RankedHit>,
	vectorHits: ReadonlyArray<RankedHit>,
): ReadonlyMap<string, RetrievalEvidence> {
	const lexical = new Map(keywordHits.map((hit) => [hit.id, hit.score]));
	const semantic = new Map(vectorHits.map((hit) => [hit.id, hit.score]));
	const evidence = new Map<string, RetrievalEvidence>();
	for (const id of new Set([...lexical.keys(), ...semantic.keys()])) {
		evidence.set(id, {
			semantic: semantic.get(id) ?? null,
			lexical: lexical.get(id) ?? null,
		});
	}
	return evidence;
}

function scoreComponents(explanation: ScoreResult): ScoreComponents {
	const rows = new Map(
		explanation.components.map((component) => [
			component.id,
			component.contribution,
		]),
	);
	const value = (id: keyof ScoreComponents): number => rows.get(id) ?? 0;
	return {
		retrieval: value("retrieval"),
		popularity: value("popularity"),
		category_match: value("category_match"),
		tag_match: value("tag_match"),
		brand_match: value("brand_match"),
		freshness: value("freshness"),
		similarity: value("similarity"),
		cooccurrence: value("cooccurrence"),
		repetition_penalty: value("repetition_penalty"),
	};
}

/**
 * Personalized score + signal breakdown for a product (scorer.score_product).
 * `weights` come from the synced bundle's ranking_config.json; they default to
 * DEFAULT_RANKING_CONFIG so call sites on an older bundle keep today's scores.
 *
 * `similarity` is the per-candidate cosine to a seed product, threaded in by the
 * vector_similarity strategies; `cooccurrence` is the per-candidate co-occurrence
 * score to a seed, threaded in by the co_occurrence strategies. Both default to 0
 * so every other path reduces to the original Phase-1 formula byte-for-byte.
 */
export function scoreProduct(
	product: Product,
	profile: SessionProfile,
	weights: ScoringWeights = DEFAULT_RANKING_CONFIG.scoring_weights,
	similarity = 0,
	cooccurrence = 0,
	retrieval = 0,
): SearchResult {
	const catMatch = profile.categoryAffinity.get(product.category) ?? 0;

	let tagMatch = 0;
	if (product.tags.length > 0) {
		const total = product.tags.reduce(
			(sum, tag) => sum + (profile.tagAffinity.get(tag) ?? 0),
			0,
		);
		tagMatch = total / product.tags.length;
	}

	const brandMatch = product.brand
		? (profile.brandAffinity.get(product.brand) ?? 0)
		: 0;

	const signals: FormulaSignals = {
		retrieval,
		popularity: product.popularity_score,
		category_match: catMatch,
		tag_match: tagMatch,
		brand_match: brandMatch,
		freshness: product.freshness_score,
		similarity,
		cooccurrence,
		repetition_penalty: Number(profile.recentlyViewed.includes(product.id)),
	};
	const explanation = explainScore(signals, weights);

	return {
		product,
		score: explanation.score,
		score_components: scoreComponents(explanation),
		score_explanation: explanation,
	};
}

/**
 * Drop every candidate that fails the absolute floor, then re-score the survivors
 * against the profile and sort descending. Mirrors reranker.rerank_search: a stable
 * descending sort (ties keep input order), matching Python's list.sort stability so
 * fused-rank ties resolve identically.
 *
 * WHY THE FLOOR RUNS FIRST, AND WHY IT DROPS RATHER THAN DEMOTES
 * `r.score` is the RRF fused score, which is built from POSITIONS: a document
 * ranked first by one retriever scores 1/61 whether it is a perfect match or the
 * least bad of 720 wrong answers. The normalization below then divides by the best
 * hit in this very result set, so the top result always takes the full relevance
 * weight — by construction, for every query ever asked. Nothing downstream can
 * express "no good match", so a full page came back for queries with no answer at
 * all. Ranking junk last does not fix that; only refusing to return it does.
 */
export function rerank(
	results: ReadonlyArray<SearchResult>,
	evidence: ReadonlyMap<string, RetrievalEvidence>,
	profile: SessionProfile,
	weights: ScoringWeights = DEFAULT_RANKING_CONFIG.scoring_weights,
): ReadonlyArray<SearchResult> {
	const admitted = results.filter((r) =>
		meetsRelevanceFloor(evidence.get(r.product.id)),
	);
	const maxRetrieval = Math.max(0, ...admitted.map((result) => result.score));
	const rescored = admitted.map((r, index) => ({
		result: scoreProduct(
			r.product,
			profile,
			weights,
			0,
			0,
			maxRetrieval === 0
				? 0
				: SEARCH_RELEVANCE_WEIGHT * (Math.max(0, r.score) / maxRetrieval),
		),
		index,
	}));
	rescored.sort((a, b) => b.result.score - a.result.score || a.index - b.index);
	return rescored.map((entry) => entry.result);
}

/**
 * Re-score every candidate threading its per-id similarity (cosine to a seed),
 * then sort descending — the strategy-aware path mirroring recommend._rerank.
 * `similarity` maps product id → cosine; absent ids score 0, so a strategy whose
 * weights.similarity is 0 reduces to the plain `rerank` result.
 */
export function rerankWithSimilarity(
	results: ReadonlyArray<SearchResult>,
	profile: SessionProfile,
	weights: ScoringWeights,
	similarity: ReadonlyMap<string, number>,
): ReadonlyArray<SearchResult> {
	const rescored = results.map((r, index) => ({
		result: scoreProduct(
			r.product,
			profile,
			weights,
			similarity.get(r.product.id) ?? 0,
		),
		index,
	}));
	rescored.sort((a, b) => b.result.score - a.result.score || a.index - b.index);
	return rescored.map((entry) => entry.result);
}

/**
 * Re-score every candidate threading its per-id co-occurrence score (the seed's
 * neighbour strength), then sort descending — the co_occurrence strategy path
 * mirroring recommend._rerank. `cooccurrence` maps product id → neighbour score;
 * absent ids score 0, so a strategy whose weights.cooccurrence is 0 reduces to the
 * plain `rerank` result. similarity is 0 on this path (co-occurrence carries no
 * cosine), so Phase-1/2 strategies stay byte-identical.
 */
export function rerankWithCooccurrence(
	results: ReadonlyArray<SearchResult>,
	profile: SessionProfile,
	weights: ScoringWeights,
	cooccurrence: ReadonlyMap<string, number>,
): ReadonlyArray<SearchResult> {
	const rescored = results.map((r, index) => ({
		result: scoreProduct(
			r.product,
			profile,
			weights,
			0,
			cooccurrence.get(r.product.id) ?? 0,
		),
		index,
	}));
	rescored.sort((a, b) => b.result.score - a.result.score || a.index - b.index);
	return rescored.map((entry) => entry.result);
}
